### DISCUSSION TITLE
`bbin` vs `bb install`

### DISCUSSION THREAD

**Discussion by rads** [babashka/bbin#6] [General]
I was inspired to create `bbin` from the discussion of the `bb install` proposal: https://github.com/babashka/babashka/discussions/1307

This thread can be used to discuss further thoughts about how `bbin` relates to the `bb install` proposal.

**Comment by rads** [comment]
Since we're using `deps.clj` for fetching dependencies (except for `:http/url`), I was able to add support for `:mvn/version` coordinates pretty easily.
```
$ bbin install org.babashka/http-server --mvn/version 0.1.11
{:lib org.babashka/http-server, :coords {:mvn/version "0.1.11"}}

$ http-server --help
Serves static assets using web server.
Options:
  * `:dir` - directory from which to serve assets
  * `:port` - port
 
$ bbin install org.babashka/http-server --as mime-types --mvn/version 0.1.11 --main-opts '["-e" "(deref (requiring-resolve (symbol \"babashka.http-server/default-mime-types\")))"]'
{:lib org.babashka/http-server, :coords {:mvn/version "0.1.11"}}

$ mime-types | jet -q '"edn" symbol'
application/edn
```

**Comment by arichiardi** [comment]
Hey there, great job on this little tool!

I might have a newbie question, does `bbin install <url>` work against any possible file, as long as you have permission?
What if you are not allowed the download? Is there a way to set headers?

We have a use case where we want to "install/deploy" a script from our internal repository (Azure DevOps) to dev machines.

Thank you in advance!

**Reply by rads** [reply]
@arichiardi: Thanks, I'm glad you like it!

> I might have a newbie question, does bbin install <url> work against any possible file, as long as you have permission?

Any local file ending in `.clj` should work, but HTTP URLs will likely fail if the response isn't 2xx (I haven't tested this myself yet).

> What if you are not allowed the download? Is there a way to set headers?

There is no way to set headers for HTTP downloads right now. That said, I think it would make sense to support this. I created an issue for it: https://github.com/babashka/bbin/issues/42

> We have a use case where we want to "install/deploy" a script from our internal repository (Azure DevOps) to dev machines.

If you have a private Git repo you want to use with `bbin` _today_, you could choose either of the following approaches (assuming you have access to Git via SSH):

1. Clone the scripts repo and install each script from the local filesystem

    ```
    $ git clone git@github.com:borkdude/tools.git
    $ bbin install tools/antq.clj
    $ bbin install tools/ls_jar.clj
    $ bbin install tools/nrepl_proxy.clj
    $ rm -rf ./tools
    ```

OR

2. Install each script from the Git repo separately (requires a `deps.edn` file in the root of the repo)
    ```
    $ bbin install io.github.borkdude/tools --git/url git@github.com:borkdude/tools.git --git/sha d4b03eb2967115134fa6aace1b2d5759a19c13e2 --main-opts '["-f" "tools/antq.clj"]' --as antq
    $ bbin install io.github.borkdude/tools --git/url git@github.com:borkdude/tools.git --git/sha d4b03eb2967115134fa6aace1b2d5759a19c13e2 --main-opts '["-f" "tools/ls_jar.clj"]' --as ls-jar
    $ bbin install io.github.borkdude/tools --git/url git@github.com:borkdude/tools.git --git/sha d4b03eb2967115134fa6aace1b2d5759a19c13e2 --main-opts '["-f" "tools/nrepl_proxy.clj"]' --as nrepl-proxy
    ```
    This approach is pretty verbose right now but it should be improved once https://github.com/babashka/bbin/issues/18 is implemented.
